### PR TITLE
Improvments to Chainwatch Syncing and Processing Pipeline

### PR DESCRIPTION
**Pull request by frrist**
### Goal
The primary goal of this change is to make iteration while developing against a real network faster. Currently chainwatch only persists blocks to the `blocks_synced` table after it has extracted and stored the information contained in their state roots. This means that every time chainwatch gets restarted (a frequent occurrence while developing) it needs to walk the blockchain up to the last block it synced. Walking the chain is a serial operation and can become quite lengthy as the chain grows. This leads to slow development velocity when working against a live network. (Working against a live network is often necessary as the data we are trying to extract only exists there).

(Github doesn't appear to follow the below links correctly if the file diff is too large :disappointed:  )

### What Changed
This PR addresses this by separating the chain-syncing-process from the data-extraction-process. 
The [`blocks_synced`](https://github.com/filecoin-project/lotus/pull/2396/files#diff-486c8cb0ee1640ddb4c2ffa3ec6fd958R53) table has been modified as:
| cid | synced_at | processed_at |
|-----|-----------|--------------|
| block cid | when block was synced | when block was processed |

When chainwatch is ran it will first start a [`Syncer`](https://github.com/filecoin-project/lotus/pull/2396/files#diff-486c8cb0ee1640ddb4c2ffa3ec6fd958R136) that continuously collects blocks from the `ChainNotify` channel and persists them to the `blocks_synced` table. Once the `Syncer` has caught the `blocks_synced` table up to the lotus daemons current head a [`Processor`](https://github.com/filecoin-project/lotus/pull/2396/files#diff-391a8b458b8546e740d7892ba2ac0d3fR79) is started. The `Processor` [selects a batch of contiguous blocks](https://github.com/filecoin-project/lotus/pull/2396/files#diff-391a8b458b8546e740d7892ba2ac0d3fR243-R253) and [extracts and stores their data](https://github.com/filecoin-project/lotus/pull/2396/files#diff-391a8b458b8546e740d7892ba2ac0d3fR91-R141). It attempts to do as much work as it can in parallel. When the blocks are done being processed the corresponding `processed_at` field in the `blocks_synced` table is filled out with a timestamp.

In ceed37f50c8a77cd55333f6e9ede7c3d2b7954d5 I removed code that appeared to longer be maintained, if this was a mistake I am happy to revert. 

### Trade-offs
This PR thrown together over the weekend to allow faster developer iteration against chainwatch, it is not a rearchitect of chainwatch. All existing tables were copy and pasted as was the code to persist data to them with a couple things [shoehorned](https://github.com/filecoin-project/lotus/pull/2396/files#diff-34152724f1e9407a19610409a3a89ea7R135) in or left [unaddressed](https://github.com/filecoin-project/lotus/pull/2396/files#diff-486c8cb0ee1640ddb4c2ffa3ec6fd958R187).

### Looking forward
- As has been stated in review, the syncer and processor both have [separate schemas](https://github.com/filecoin-project/lotus/pull/2396#discussion_r453923994), this needs more design.
- Genesis state [needs special handling implemented](https://github.com/filecoin-project/lotus/pull/2396/files#diff-391a8b458b8546e740d7892ba2ac0d3fR96-R97)
- Syncer and Processor need more accurate names
- Godocs should be added once we stabilize the abstractions.
- TESTING!
  - Looks like a good database mock: https://github.com/cockroachdb/copyist
- Chainwatch metrics
  - Duration of difference processes
  - CPU and Memory usage, etc.
- Probably other things I am forgetting. 

Contributions for https://github.com/filecoin-project/sentinel/issues/24

### CODE REVIEW

File: cmd/lotus-chainwatch/processor/miner.go [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+	if err := p.updateMiners(ctx, minerChanges); err != nil {
+		log.Fatalw("Failed to update miner actors", "error", err)
+	}
+	return nil
+}
+
+func (p *Processor) processMiners(ctx context.Context, minerTips map[types.TipSetKey][]actorInfo) (out []minerActorInfo, err error) {
+	start := time.Now()
+	defer func() {
+		if err == nil {
```
Comment (placer14) [inline cmd/lotus-chainwatch/processor/miner.go:166]: This `err` does not get assigned anywhere below and will always be nil.

Comment (placer14) [reply cmd/lotus-chainwatch/processor/miner.go:166]: You're using the return err value, but they are set to `nil` on creation and never changed. (The closures below all use `:=` and will create a new instance of `err` inside and then get destroyed at the end of the outermost `for`.)

File: cmd/lotus-chainwatch/processor/mpool.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	loop:
+		for {
+			time.Sleep(10 * time.Millisecond)
+			select {
+			case update := <-sub:
+				updates = append(updates, update)
+			default:
+				break loop
+			}
+		}
```
Comment (placer14) [inline cmd/lotus-chainwatch/processor/mpool.go:40]: This looks like you're trying to make the loop run on a timer? What's the motivation with these loops?

File: cmd/lotus-chainwatch/processor/processor.go [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+			if err := p.refreshViews(); err != nil {
+				log.Errorw("Failed to refresh views", "error", err)
+			}
+		}
+	}()
+
+}
+
+func (p *Processor) refreshViews() error {
+	if _, err := p.db.Exec(`refresh materialized view state_heights`); err != nil {
```
Comment (placer14) [inline cmd/lotus-chainwatch/processor/processor.go:141]: Let's run this with `CONCURRENTLY` which leaves the table available for query while it's updating. This requires that we add a `UNIQUE` index on `state_heights` which can be on the `parentstateroot` column.
```suggestion
	if _, err := p.db.Exec(`refresh materialized view concurrently state_heights`); err != nil {
```

File: cmd/lotus-chainwatch/processor/processor.go [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+func (p *Processor) unprocessedBlocks(ctx context.Context, batch int) (_ map[cid.Cid]*types.BlockHeader, err error) {
+	start := time.Now()
+	defer func() {
+		if err == nil {
+			log.Infow("Gathered Blocks to process", "duration", time.Since(start).String())
+		}
+	}()
+	rws, err := p.db.Query(`
```
Comment (placer14) [inline cmd/lotus-chainwatch/processor/processor.go:236]: Don't forget to `rws.Close()`

File: cmd/lotus-chainwatch/processor/processor.go [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+		if err != nil {
+			return nil, xerrors.Errorf("Failed to parse unprocessed blocks: %w", err)
+		}
+		bh, err := p.node.ChainGetBlock(ctx, ci)
+		if err != nil {
+			// this is a pretty serious issue.
+			return nil, xerrors.Errorf("Failed to get block header %s: %w", ci.String(), err)
+		}
+		out[ci] = bh
+	}
```
Comment (placer14) [inline cmd/lotus-chainwatch/processor/processor.go:268]: Calling rws.Next() might error. Make sure you check for it:
```
	if err := rws.Err(); err != nil {
		log.Fatal(err)
	}
```

Also.... rws? Was the `o` so expensive?

File: cmd/lotus-chainwatch/syncer/sync.go [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+}
+
+func NewSyncer(db *sql.DB, node api.FullNode) *Syncer {
+	return &Syncer{
+		db:   db,
+		node: node,
+	}
+}
+
+func (s *Syncer) setup() error {
```
Comment (placer14) [inline cmd/lotus-chainwatch/syncer/sync.go:36]:
```suggestion
func (s *Syncer) setupSchema() error {
```

File: cmd/lotus-chainwatch/syncer/sync.go [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+
+	headerLk sync.Mutex
+	node     api.FullNode
+}
+
+func NewSyncer(db *sql.DB, node api.FullNode) *Syncer {
+	return &Syncer{
+		db:   db,
+		node: node,
+	}
```
Comment (placer14) [inline cmd/lotus-chainwatch/syncer/sync.go:33]: Consider checking the db state to confirm it's valid with `err := db.Ping()`.

File: cmd/lotus-chainwatch/processor/mpool.go (hunk 2 above)
Comment (frrist) [reply cmd/lotus-chainwatch/processor/mpool.go:40]: This is copy-pasted from https://github.com/filecoin-project/lotus/blob/3f017688c7ae180d12d4ee28f732298299ee351f/cmd/lotus-chainwatch/mpool.go.

File: cmd/lotus-chainwatch/main.go [hunk 8]
```diff
@@ -85,22 +85,14 @@ var runCmd = &cli.Command{
...
-
-		h, err := newHandler(api, st)
-		if err != nil {
-			return xerrors.Errorf("handler setup: %w", err)
-		}
+		db, err := sql.Open("postgres", cctx.String("db"))
+		defer db.Close()
 
-		http.Handle("/", h)
+		sync := syncer.NewSyncer(db, api)
```
[the remark this answers is not in the document]
Comment (frrist) [reply cmd/lotus-chainwatch/main.go:91]: I'll just do it in main and bail if it doesn't respond.

File: cmd/lotus-chainwatch/processor/miner.go [hunk 9]
```diff
@@ -0,0 +1,655 @@
...
+create table if not exists miner_sector_events
+(
+    miner_id text not null,
+    sector_id bigint not null,
+    state_root text not null,
+    event miner_sector_event_type not null,
+    
+	constraint miner_sector_events_pk
+		primary key (sector_id, event, miner_id, state_root)
+)
```
[the remark this answers is not in the document]
Comment (frrist) [reply cmd/lotus-chainwatch/processor/miner.go:117]: Yeah, just copy/pasta.

File: cmd/lotus-chainwatch/processor/miner.go (hunk 1 above)
Comment (placer14) [reply cmd/lotus-chainwatch/processor/miner.go:166]: Disregard. I guess my local testing didn't exercise your pattern exactly the same. This appears to work as intended.

File: cmd/lotus-chainwatch/processor/processor.go [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+
+	if err := p.setupSchemas(); err != nil {
+		log.Fatalw("Failed to setup processor", "error", err)
+	}
+
+	go p.subMpool(ctx)
+
+	// main processor loop
+	go func() {
+		for {
```
Comment (placer14) [inline cmd/lotus-chainwatch/processor/processor.go:90]: This `for` should handle ctx.Done and exit.

File: cmd/lotus-chainwatch/processor/processor.go [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+	start := time.Now()
+	defer func() {
+		log.Infow("Gathered Blocks to process", "duration", time.Since(start).String())
+	}()
+	rows, err := p.db.Query(`
+with toProcess as (
+    select blocks.cid, blocks.height, rank() over (order by height) as rnk
+    from blocks
+        left join blocks_synced bs on blocks.cid = bs.cid
+    where bs.is_processed is false and blocks.height != 0
```
Comment (placer14) [inline cmd/lotus-chainwatch/processor/processor.go:248]: A slightly better condition which excludes the possibility of negative ints and NULLs.
```suggestion
    where bs.is_processed is false and blocks.height > 0
```

Comment (placer14) [reply cmd/lotus-chainwatch/processor/processor.go:248]: Also, I saw you mentioned `processed_at` and `is_processed` are both included in this table. Consider using only `processed_at` which can be NULL when is_processed = false.

File: cmd/lotus-chainwatch/processor/processor.go (hunk 10 above)
Comment (frrist) [reply cmd/lotus-chainwatch/processor/processor.go:90]: In `main.go` we (somewhat ungracefully) call `os.Exit()` when `ctx.Done()` returns. Would there be any value added by implementing a graceful termination here in light of this?

Comment (placer14) [reply cmd/lotus-chainwatch/processor/processor.go:90]: If we somehow change the app in a way that the context ends but the app continues running, this goroutine will never resolve. Someone who's moving the consumption of processer.Start shouldn't have to know there's an immortal goroutine living within that now needs to be handled properly. They see `Start(ctx context.Context)` and assume the context is respected (and rightfully so).

**Comment by frrist** [comment]
Will rebase and add support for features added in https://github.com/filecoin-project/lotus/pull/2331

**Comment by placer14** [comment]
FYI @frrist `make chainwatch` is broken on this branch.

**Comment by frrist** [comment]
@placer14 thanks for flagging, should be good now.

**Pull request merged**